Various bugfixes

smathot opened the pull request:
@dschreij I will append various minor bugfix commits to this PR, for you to review when you get back :-)

smathot committed:
Correctly parse variably defined timeouts (closes #27)

smathot committed:
Convert newlines to <br /> tags (closes #20)

smathot committed:
Properly reset loop table after break-if (closes #30)

smathot committed:
loop: perform loop operations at the end of creating the live matrix (closes #23)

smathot committed:
sequence: strip slashes and remove quotes from conditional statement (closes #24)

smathot committed:
Fix repeat_cycle handling (closes #35)

smathot committed:
Fix break-if statement (closes #31)

smathot committed:
Parse nested variable definitions (closes #17)

smathot committed:
Fix advanced_delay when duration is 0 (closes #36)

dschreij commented on src/js/osweb/backends/canvas.js:
```diff
@@ -893,6 +893,9 @@ export default class Canvas {
   text (txt, center, x, y, html, styleArgs) {
+    // Newlines are not dealt with correctly, so we convert them to <br />
+    // tags, which results in the same behavior as the desktop.
+    txt = txt.replace('\n', '<br />')
```
The space between r and / is not strictly required, but also not a problem to leave in.

dschreij commented on src/js/osweb/items/loop.js:
```diff
@@ -79,15 +78,14 @@ export default class Loop extends Item {
   /** Reset all item variables to their default value. */
   reset () {
-    this.matrix = []
+    this.orig_matrix = []
```
It is really hard to follow what you are doing here, or why the orig_matrix is required to make resumptions possible, but I assume you tested it and this works. If so, I am ok with it.

smathot replied:
The `orig_matrix` is the original state of the matrix before any of the operations are applied. And then the `matrix` is the matrix as it used during a particular run of the `loop` item. This allows the loop item to run multiple times from the same state.

dschreij commented on src/js/osweb/items/loop.js:
```diff
@@ -134,52 +132,41 @@ export default class Loop extends Item {
             break
           case 'fullfactorial':
             this._operations.push([fullfactorial, []])
-            // this.matrix = fullfactorial(this.matrix)
```
Oops, I guess I forgot to clean up after myself again ;)

dschreij commented on src/js/osweb/classes/syntax.js:
```diff
@@ -153,18 +148,27 @@ export default class Syntax {
-    return this.strip_slashes(this.unescapeBrackets(result))
+    return this.strip_slashes(this.unescapeBrackets(text))
   }
 
```
Nice! This is also easier to follow this way

smathot committed:
Parse multiline variables
- Closes #37
- Makes hacky implementatin of inline_script.from_string() redundant

smathot commented:
Sorry, I didn't notice that you had approved the PR so I appended c2b84c35c6da332c6e9a69460c6388e17b057b01. Could you quickly review that one too? And then if you approve I suggest that we release 1.3.9, and at the same time release OpenSesame 3.3.4 with *many* bugfixes!

dschreij commented on src/js/osweb/classes/syntax.js:
```diff
@@ -222,6 +222,39 @@ export default class Syntax {
+    let dedented = []
+    for (const line of lines) {
+      if (line[0] !== '\t') {
+        return script
```
This seems peculiar. If the first character of a line is not a tab, return the script (e.g. the input variable?) so for instance a script like
```
    define something
    do something else
non-indented line
```
would just return the script since the last line does not start with a tab, even though you might still want to dedent the first two lines?
I assume this is just my lack of understanding the situation and you have tested this (judging from the unit tests) so it just works.

smathot replied:
Yes, that's intended. Basically, it should dedent a script that corresponds to an item definition, and this is always indented by a single tab. If any line is not indented, then it's not an item definition, and it should be left as is.

dschreij commented on src/js/osweb/items/inline_script.js:
```diff
@@ -56,54 +56,6 @@ export default class InlineScript extends Item {
     this.vars._run = ''
   }
 
-  /**
```
inline_script is not supported anyway, right? Should we leave this file in at all?
